tidy3d/components/geometry/polyslab.py

```diff
@@ -552,6 +553,8 @@ def intersections_tilted_plane(
             For more details refer to
             `Shapely's Documentation <https://shapely.readthedocs.io/en/stable/project.html>`_.
         """
+        import trimesh
```

Comment:
I was dealing with a similar thing in my 3.0 architecture exercise yesterday. I was wondering if there's any way to "inject" this package into the function scope namespace, such that if `verify_packages_import("x")` works, then `x` is imported in the function's scope (and we dont need to import it manually). sounds like it could be potentially buggy but just a thought.

Reply:
This is actually really interesting - because I've been thinking about how best to do this in tidy3d 3.0 too, and it doesn't involve on-the-fly imports not necessarily like this format style.

There has been some debate around this here https://stackoverflow.com/questions/128478/should-import-statements-always-be-at-the-top-of-a-module and I think in our case it's fine. Re -importing the module within the function is still fine to solve the namespace situation and probably cleaner than the alternative variable as the memory of the imported library is managed by importlib rather than us

Reply:
if the import has already been imported in another method, the 2nd import only takes minimal amount of time right? (unless we manually reload it with one of import lib's functions?)

I feel like fundamentally the problem here is that we're trying to mix optional dependencies within the core components, when they should probably be plugins.. although that could hurt the user experience a bit.

Reply:
That's the idea, it's not great but working within the current structure.

Yeah I agree, I've been trying to keep them separate in different files and only use methods from those files within any core methods if needed. But really our architecture should be modular like the plugins in this regard I think too.

Reply:
One possible solution to avoid calling `import` within the function is to get the imported module from `sys.modules` (it might be more efficient as well, since we know the module is already imported):
```python
In [1]: import importlib

In [2]: import sys

In [3]: def check_import(s):
   ...:     try:
   ...:         importlib.import_module(s)
   ...:         return True
   ...:     except ImportError:
   ...:         return False
   ...:

In [4]: def f():
   ...:     if not check_import("numpy"):
   ...:         raise ImportError("numpy not available")
   ...:     # avoid import here
   ...:     numpy = sys.modules["numpy"]
   ...:     return numpy.arange(3)
   ...:

In [5]: f()
Out[5]: array([0, 1, 2])
```

Reply:
I was interested about this, turns out this is exactly what happens under the hood when we do `import_module`:

https://github.com/python/cpython/blob/c63c6142f9146e1e977f4c824c56e8979e6aca87/Lib/importlib/_bootstrap.py#L1354